### PR TITLE
Added a nullcheck for username and an appropriate exception to avoid …

### PR DESCRIPTION
**Pull request by shindouj**
…a NullPointerException when username is null

Signed-off-by: Jakub Strzyżewski <github@jeikobu.net>

### Context
Fix for issue #6369; PR requested by @marcphilipp  

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective [Change too trivial to warrant an integration test; functionality checked in unit test]
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes [No Javadocs exist for modified class (internal package)]
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/ntlm/NTLMCredentials.java [hunk 1]
```diff
@@ -32,6 +32,11 @@
     public NTLMCredentials(PasswordCredentials credentials) {
         String domain;
         String username = credentials.getUsername();
+
+        if (username == null) {
+            throw new IllegalArgumentException("Username must not be null!");
+        }
+
```
Comment (marcphilipp) [inline subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/ntlm/NTLMCredentials.java]: On second thought, please use `Preconditions.checkNotNull` which you can use like this:


```java
String username = Preconditions.checkNotNull(credentials.getUsername(), "username must not be null");
```

**Comment by shindouj** [comment]
I've made the changes you (@marcphilipp) requested, but I don't think they are fixing the issue at all.

Adding a message to a very complicated stack trace isn't helping. The core issue is that the error message while the username is null is as follows:

`* What went wrong:
Execution failed for task ':publishMavenJavaPublicationToMavenRepository'.
> Failed to publish publication 'mavenJava' to repository 'maven'
   > Failed to retrieve remote metadata net.jeikobu:jbase:1.0-SNAPSHOT/maven-metadata.xml: Could not transfer metadata net.jeikobu:jbase:1.0-SNAPSHOT/maven-metadata.xml from/to remote (https://nexus.jeikobu.net/repository/maven-snapshots/): Could not get resource 'net/jeikobu/jbase/1.0-SNAPSHOT/maven-metadata.xml'`

The nullcheck should be either propagated to this level or be made earlier in the process, so the message "Username should not be null" is either replacing the above message or concatenated to it, and that's because not being able to get that resource is not really a failing issue for publishing with maven-publish. 

Additionally I was forced to remove the new unit test because it would mean checking for NullPointerException.

**Comment by marcphilipp** [comment]
> I've made the changes you requested, but after additional tests I don't think your suggested fixes are addressing the core issue at all.

I agree that it's not enough, but it at least provides a better message when using `--stacktrace`. In addition, you're right, we should validate the username as early as possible.

> Additionally I was forced to remove the new unit test because it would mean checking for NullPointerException.

Please add it again and verify the exception message as well.

**Comment by shindouj** [comment]
@marcphilipp

**Comment by marcphilipp** [comment]
Successful build: https://builds.gradle.org/viewLog.html?buildId=15480212&buildTypeId=Gradle_Check_Stage_BranchBuildAccept_Trigger

**Pull request merged**